File: src/maya_usd_export/interface/widgets/collapsible_container.py
from maya_usd_export.utils import pyside_importer
_, qtc, qtw, shiboken = pyside_importer.import_all()

class CollapsibleContainerHeader(qtw.QPushButton):
    def __init__(self, title):
        super().__init__(title)
        main_layout = qtw.QVBoxLayout()
        self.setLayout(main_layout)

class CollapsibleContainer(qtw.QWidget):
    def __init__(self, title):
        super().__init__()
        self._main_layout = qtw.QVBoxLayout()
        self._main_layout.setAlignment(qtc.Qt.AlignTop)
        self.setLayout(self._main_layout)

        self._head_bttn = CollapsibleContainerHeader(title)
        self._head_bttn.clicked.connect(self.header_clicked_slot)
        self._head_bttn.setIcon(self.style().standardIcon(qtw.QStyle.SP_ArrowDown))

        self._main_layout.addWidget(self._head_bttn)

        self._contents_w = qtw.QWidget()
        self._main_layout.addWidget(self._contents_w)
        self._contents_l = qtw.QVBoxLayout(self._contents_w)
        self._contents_l.setContentsMargins(0,0,0,0)

    def header_clicked_slot(self):
        self._contents_w.setVisible(not self._contents_w.isVisible())
        if(self._contents_w.isVisible()):
            self._head_bttn.setIcon(self.style().standardIcon(qtw.QStyle.SP_ArrowDown))
        else:
            self._head_bttn.setIcon(self.style().standardIcon(qtw.QStyle.SP_ArrowRight))

    def addWidget(self, _widget):
       self._contents_l.addWidget(_widget) 

    def addLayout(self, _layout):
       self._contents_l.addLayout(_layout) 
        

File: src/interface/widgets/CollapsibleContainer.cpp
#include "interface/widgets/CollapsibleContainer.h"
#include "QtCore/qnamespace.h"
#include "QtWidgets/qboxlayout.h"
#include "QtWidgets/qpushbutton.h"
#include "QtWidgets/qwidget.h"
#include <iostream>

#include <QtWidgets/QVBoxLayout>
#include <QtWidgets/QPushButton>


CollapsibleContainer::CollapsibleContainer(const QString& _title, QWidget* _parent)
{
    m_mainLayout = new QVBoxLayout();
    m_mainLayout->setAlignment(Qt::AlignTop);
    this->setLayout(m_mainLayout);

    m_headButton = new QPushButton("    "+_title);
    QObject::connect(m_headButton, &QPushButton::clicked, this, &CollapsibleContainer::headerClickedSlot);
    m_headButton->setIcon(QIcon(":/icons/collapse_arrow_down.png"));
    m_headButton->setStyleSheet("text-align:left;");


    m_mainLayout->addWidget(m_headButton);

    m_contentsWidget = new QWidget();
    m_mainLayout->addWidget(m_contentsWidget);
    m_contentsLayout = new QVBoxLayout(m_contentsWidget);
    m_contentsLayout->setContentsMargins(0,0,0,0);
}

void CollapsibleContainer::addWidget(QWidget *_widget)
{
    m_contentsLayout->addWidget(_widget);
}
void CollapsibleContainer::addLayout(QLayout *_layout)
{
    m_contentsLayout->addLayout(_layout);
}

void CollapsibleContainer::headerClickedSlot()
{

    m_contentsWidget->setVisible(! m_contentsWidget->isVisible());
    if(m_contentsWidget->isVisible())
    {
        // TODO: set icon down
        m_headButton->setIcon(QIcon(":/icons/collapse_arrow_down.png"));
    }
    else
    {
        // TODO: set icon right
        m_headButton->setIcon(QIcon(":/icons/collapse_arrow_right.png"));
    }
}
